sdk/nodejs/cmd/run-plugin/index.ts

```diff
@@ -0,0 +1,127 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The very first thing we do is set up unhandled exception and rejection hooks to ensure that these
+// events cause us to exit with a non-zero code. It is critically important that we do this early:
+// if we do not, unhandled rejections in particular may cause us to exit with a 0 exit code, which
+// will trick the engine into thinking that the program ran successfully. This can cause the engine
+// to decide to delete all of a stack's resources.
+//
+// We track all uncaught errors here.  If we have any, we will make sure we always have a non-0 exit
+// code.
+const uncaughtErrors = new Set<Error>();
+
+// We also track errors we know were logged to the user using our standard `log.error` call from
+// inside our uncaught-error-handler in run.ts.  If all uncaught-errors above were also known to all
+// be logged properly to the user, then we know the user has the information they need to proceed.
+// We can then report the langhost that it should just stop running immediately and not print any
+// additional superfluous information.
+const loggedErrors = new Set<Error>();
+
+let programRunning = false;
+const uncaughtHandler = (err: Error) => {
+    uncaughtErrors.add(err);
+    if (!programRunning) {
+        console.error(err.stack || err.message || "" + err);
+    }
+};
+
+// Keep track if we already logged the information about an unhandled error to the user..  If
+// so, we end with a different exit code.  The language host recognizes this and will not print
+// any further messages to the user since we already took care of it.
+//
+// 32 was picked so as to be very unlikely to collide with any of the error codes documented by
+// nodejs here:
+// https://github.com/nodejs/node-v0.x-archive/blob/master/doc/api/process.markdown#exit-codes
```

Comment:
I know this is copy-pasted, but I suspect https://nodejs.org/api/process.html#process_exit_codes is a better link for this content these days.

Reply:
Nice, I'll update it in other places as well.